### PR TITLE
Elevation per polygon for fill extrusions

### PR DESCRIPTION
**Pull request by antonmarsden**
This PR is a basic solution to [issue #3313](https://github.com/maplibre/maplibre-gl-js/issues/3313), and renders multipolygons at their own centroid's elevation.

It is potentially a breaking change for some edge cases though, and it may be worth introducing a new **fill-extrusion** style property (e.g., ""fill-extrusion-elevation-per-polygon") to optionally activate this behaviour. Suggestions welcomed :)

 - [X] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [X] Briefly describe the changes in this PR.
 - [X] Link to related issues.
 - [X] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality.
 - [X] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [X] Add an entry to `CHANGELOG.md` under the `## main` section.

**Before:**
before

**After:**
after

### CODE REVIEW

**Comment by HarelM** [comment]
Thanks for taking the time to look into this!

This is only relevant for terrain as it will be kept at 0 for non terrain maps, right?
If my above assumption is correct I don't think there's a need to change the style spec for that.
Can you add a render test that demonstrates this issue/solution?
Also, in the linked bug, there's a jump in the elevation between two zoom levels (the buildings starts floating), did you happen to try and solve that as well?

**Comment by HarelM** [comment]
Also worth making sure is that buildings with multiple parts are presented correctly after this fix.
We had an issue in the past with terrain and second level building porch etc.

**Comment by acalcutt** [comment]
I just want to mention I was talking to someone in slack a while back that wanted to keep the heights of separate polygons the same when using terrain ( https://osmus.slack.com/archives/C01G3D28DAB/p1705407931974739 ) . I had suggested they could possibly use the Multipolygon grouping to kind of do what they wanted, but the fix we are trying to use here would break that option for them. If you did add an option like the 'fill-extrusion-elevation-per-polygon' it could possibly still allow that kind of use case.

**Comment by HarelM** [comment]
I think that the default should be that every polygon would be placed in the right place.
If someone needs something "special" then a design proposal in the spec and a PR with implementation can go a long way :-)

**Comment by acalcutt** [comment]
Makes sense to me and it would fix my original issue https://github.com/maplibre/maplibre-gl-js/issues/3313 , just wanted to bring up there may be some cases where this isn't wanted, like we talked about around https://github.com/maplibre/maplibre-gl-js/issues/3313#issuecomment-1803598318

I think it makes sense that the change in this PR is the default.

**Comment by HarelM** [comment]
@antonmarsden Can you please add a render test so I can merge this?

**Comment by antonmarsden** [comment]
@HarelM I have added a basic test to verify that centroids are created on a per-polygon basis. Have not had much spare time recently, am hoping this is sufficient to move things forward.

**Comment by HarelM** [comment]
Thanks! I was hoping for a render test, something similar to the screenshot you added, but hopefully simpler.
Unit test is great but less powerful in this case I think.

**Comment by antonmarsden** [comment]
@HarelM I assume you mean adding to `integration/render/tests`?

**Comment by HarelM** [comment]
Yes, create a style with a few building in a multipolygon and use a terrain from other tests to show that the buildings don't have the same elevation.

File: package.json [hunk 1]
```diff
@@ -99,6 +99,7 @@
     "is-builtin-module": "^3.2.1",
     "jest": "^29.7.0",
     "jest-environment-jsdom": "^29.7.0",
+    "jest-mock-extended": "^3.0.5",
```
Comment (HarelM) [inline package.json:102]: Is this really needed? I believe jest has a built in capabilities of mocking classes, doesn't it?

**Comment by HarelM** [comment]
CC: @louwers note for parity. Should be excluded as it used terrain.

**Comment by louwers** [comment]
Can you move the render tests under `terrain`?

**Review by louwers — DISMISSED** [review]
Can you move the render tests under `terrain` since it is related to (and requires) this feature? Then it won't show up as a disparity.

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,12 +2,16 @@
 
 ### ✨ Features and improvements
 
+- Improve animation curve when easeTo and flyTo with constraints ([#3793](https://github.com/maplibre/maplibre-gl-js/pull/3793))
+- For filled extrusions, calculate the elevation per polygon (see issue [#3313](https://github.com/maplibre/maplibre-gl-js/issues/3313))
+
```
Comment (HarelM) [inline CHANGELOG.md:7]:
```suggestion
```

**Review by HarelM — APPROVED** [review]

**Review by louwers — APPROVED** [review]

**Comment by louwers** [comment]
Thanks!

**Comment by petrsloup** [comment]
@HarelM any plans for a bugfix release soon? We would like to get this fix to production as soon as possible...

**Comment by HarelM** [comment]
Sure, I'll push a release in the next few days.

**Pull request merged**